### PR TITLE
[Add]: TensorRT Support

### PR DESCRIPTION
**Pull request by gpastal24**
Adding tensorrt support for HigherHRNet

### CODE REVIEW

File: SimpleHigherHRNet.py [hunk 1]
```diff
@@ -30,7 +106,8 @@ def __init__(self,
                  filter_redundant_poses=True,
                  max_nof_people=30,
                  max_batch_size=32,
-                 device=torch.device("cpu")):
+                 device=torch.device("cpu"),
+                 trt_=False):
```
Comment (stefanopini) [inline SimpleHigherHRNet.py]: I would use a more explicit name, e.g. `enable_tensorrt`.
The variable should also be added to the docstring.

File: SimpleHigherHRNet.py [hunk 2]
```diff
@@ -90,33 +168,62 @@ def __init__(self,
...
+            self.model.eval()
         else:
-            raise ValueError('Wrong device name.')
-
-        self.model = self.model.to(device)
-        self.model.eval()
+            # import pycuda.driver as cuda
+            # self.model = TrtModel('pose_higher_hrnet_w32_512.engine')
+            if device.type == 'cpu':
+                    device = torch.device('cuda:0')
```
Comment (stefanopini) [inline SimpleHigherHRNet.py]: I think this should be
```suggestion
                raise ValueError('TensorRT does not support cpu device.')
```

File: scripts/live-demo.py [hunk 3]
```diff
@@ -167,6 +173,7 @@ def main(camera_id, filename, hrnet_c, hrnet_j, hrnet_weights, hrnet_joints_set,
                                          "set to `cuda:IDS` to use one or more specific GPUs "
                                          "(e.g. `cuda:0` `cuda:1,2`); "
                                          "set to `cpu` to run on cpu.", type=str, default=None)
+    parser.add_argument("--trt_",action='store_true')
```
Comment (stefanopini) [inline scripts/live-demo.py]: Same here, more explicit name and param description.

File: scripts/live-demo.py [hunk 4]
```diff
@@ -145,7 +151,7 @@ def main(camera_id, filename, hrnet_c, hrnet_j, hrnet_weights, hrnet_joints_set,
                                                 "resnet size (if model is PoseResNet)", type=int, default=32)
     parser.add_argument("--hrnet_j", "-j", help="hrnet parameters - number of joints", type=int, default=17)
     parser.add_argument("--hrnet_weights", "-w", help="hrnet parameters - path to the pretrained weights",
-                        type=str, default="./weights/pose_higher_hrnet_w32_512.pth")
+                        type=str, default="./pose_higher_hrnet_w32_512.pth")
```
Comment (stefanopini) [inline scripts/live-demo.py:154]: The default weight path should not change

File: misc/utils.py [hunk 5]
```diff
@@ -593,3 +605,105 @@ def get_final_preds(grouped_joints, center, scale, heatmap_size):
...
+    elif dtype == trt.int32:
+        return torch.int32
+    elif dtype == trt.float16:
+        return torch.float16
+    elif dtype == trt.float32:
+        return torch.float32
+    else:
+        raise TypeError("%s is not supported by torch" % dtype)
+class TRTModule_hrnet(torch.nn.Module):
+    def __init__(self, engine=None, input_names=None, output_names=None, input_flattener=None, output_flattener=None,path=None,device=None):
```
Comment (stefanopini) [inline misc/utils.py]: Missing docstring for the class/__init__ method

File: export.py [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+    parser.add_argument('--agnostic-nms', action='store_true', help='TF: add agnostic NMS to model')
+    parser.add_argument('--topk-per-class', type=int, default=100, help='TF.js NMS: topk per class to keep')
+    parser.add_argument('--topk-all', type=int, default=100, help='TF.js NMS: topk for all classes to keep')
+    parser.add_argument('--iou-thres', type=float, default=0.45, help='TF.js NMS: IoU threshold')
+    parser.add_argument('--conf-thres', type=float, default=0.25, help='TF.js NMS: confidence threshold')
+    parser.add_argument(
+        '--include',
+        nargs='+',
+        default=['torchscript'],
+        help='torchscript, onnx, openvino, engine, coreml, saved_model, pb, tflite, edgetpu, tfjs, paddle')
```
Comment (stefanopini) [inline export.py]: Are all these params needed? Some of them seem to be relevant to yolov5 only

**Review by stefanopini — CHANGES_REQUESTED** [review]
Hi @gpastal24 , thanks for the great contribution!!

The code looks good to me in general (still have to try running it though).

There are a few minor things to fix before merging:
- there is a lot of commented code that seems old/debug code, it should be removed
- some docstrings of the main functions are missing
- `__pycache__` files shouldn't be committed
- I would put all the utils from yolov5 in a subfolder of utils, e.g. `utils/yolov5`
- I would split the export.py file in two parts:
  - `utils/yolov5/export.py` with all the functions that have remained untouched from the original yolov5 repo
  - `scripts/export_model.py` with the actual functions that have changed to work with SimplerHRNet, e.g. including `main`, `parse_opt`, `run`
- other comments part of the review

I will create a PR to update your PR in the next few days. Then I'll ask your review and if everything looks good, we can merge it into your PR and then yours into the master branch.
Alternatively, please feel free to make changes to your PR directly to address my comments! I can review the changes when they're ready.

**Comment by gpastal24** [comment]
Hi, thanks for your time. I think most of your comments have been addressed.
If you wish to test this install nvidia-pyindex and tensorrt

` pip3 install nvidia-pyindex nvidia-tensorrt==7.2.3.4`

Trt 7 produces better results for me, but maybe this has to do with different cudnn versions between torch and trt.

This should produce the fp16 engine.

` python3 scripts/export_model.py --device 0 --half` 

And for testing.

` python3 scripts/live-demo.py --enable_tensorrt --filename=<> --hrnet_weights='pose_higher_hrnet_w32_512.engine'`

**Comment by stefanopini** [comment]
Hi @gpastal24 , I tested your code on Google Colab with CUDA 11/TensorRT 8.5 and it worked smoothly! Thanks a lot for the great contribution! ❤️ 

The code has still some minor style issues so I'm going to merge it into a new branch 'tensorrt', make minor changes to integrate/align it with the rest of the repo and then port the commits into the master branch.
Thanks again, I'll keep you posted on the next steps!

**Pull request merged**